### PR TITLE
[MGBOBR-6] Part 1: Manager-side API for adding Processor to a Bridge

### PR DESCRIPTION
**Pull request by robpblake**
This PR adds an additional endpoint to our API to add support for adding a Processor to a Bridge. The supported endpoint is `/api/v1/bridges/{id}/processors`. A Processor can be created on the given Bridge instance by `POST`-ing a request to this endpoint.

The shard side of the implementation is deliberately excluded from this PR. That will be in a follow-on PR.

Other minor changes in this PR include:

- Ensuring we return a HTTP 201 Created Response code from our API in response to POST requests as per guidelines

### CODE REVIEW

File: manager/src/main/java/com/redhat/developer/manager/ProcessorService.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        Processor p = processorDAO.findByBridgeIdAndName(bridgeId, processorRequest.getName());
+        if (p != null) {
+            throw new AlreadyExistingItemException("Processor with name '" + processorRequest.getName() + "' already exists for bridge with id '" + bridgeId + "' for customer '" + customerId + "'");
+        }
+
+        p = new Processor();
+        p.setName(processorRequest.getName());
+        p.setSubmittedAt(ZonedDateTime.now());
+        p.setStatus(BridgeStatus.REQUESTED);
+        bridge.addProcessor(p);
```
Comment (r00ta) [inline manager/src/main/java/com/redhat/developer/manager/ProcessorService.java:49]:
```suggestion
```
According to our schema definition, I think you don't need to do this

**Review by r00ta — COMMENTED** [review]
Hi @robpblake , few minor comments from my side

File: manager/src/main/java/com/redhat/developer/manager/models/Bridge.java [hunk 2]
```diff
@@ -53,9 +58,22 @@
...
+    @OneToMany(fetch = FetchType.LAZY, mappedBy = "bridge", orphanRemoval = true, cascade = CascadeType.ALL)
+    private List<Processor> processors = new ArrayList<>();
+
     public Bridge() {
     }
 
+    public void addProcessor(Processor processor) {
+        this.processors.add(processor);
+        processor.setBridge(this);
+    }
```
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/main/java/com/redhat/developer/manager/models/Bridge.java:70]: @r00ta It's a generally recognised pattern for managing relationships in `@OneToMany` in JPA. Check out : https://vladmihalcea.com/the-best-way-to-map-a-onetomany-association-with-jpa-and-hibernate/, particulary the section titled `BiDirectional @OneToMany`

File: manager/src/main/java/com/redhat/developer/manager/api/models/responses/ProcessorResponse.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    private final String kind = "Processor";
+
+    @JsonProperty("href")
+    private String href;
+
+    @JsonProperty("name")
+    private String name;
+
+    @JsonProperty("bridge")
+    private BridgeResponse bridge;
```
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/main/java/com/redhat/developer/manager/api/models/responses/ProcessorResponse.java:26]: Felt useful to be able to see details of the Bridge a Processor is attached to without having to make another request to the API.

File: manager/src/main/java/com/redhat/developer/manager/ProcessorService.java (hunk 1 above)
Comment (robpblake) [reply manager/src/main/java/com/redhat/developer/manager/ProcessorService.java:49]: This is setting up the foreign key relationship between Bridge --> Processor as part of the data i.e. a Processor cannot be inserted into the DB without being associated with a valid Bridge.

File: manager/src/test/java/com/redhat/developer/manager/dao/ProcessorDAOTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        Processor p = createProcessor(b, "foo");
+
+        Processor byBridgeIdAndName = processorDAO.findByBridgeIdAndName(b.getId(), p.getName());
+        assertThat(byBridgeIdAndName, is(notNullValue()));
+        assertThat(byBridgeIdAndName.getName(), equalTo(p.getName()));
+        assertThat(byBridgeIdAndName.getBridge().getId(), equalTo(b.getId()));
+    }
+
+    @Test
+    @Transactional
```
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/test/java/com/redhat/developer/manager/dao/ProcessorDAOTest.java:95]: I need the `@Transactional` here for the `processorDAO.getEntityManager.merge()` calls as there is no active transaction otherwise.

File: manager/src/main/java/com/redhat/developer/manager/models/Bridge.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielezonca) [reply manager/src/main/java/com/redhat/developer/manager/models/Bridge.java:70]: I think the name of the method is a bit misleading: I usually don't expect a `add...(value)` to perform a side effect on the argument `value`.
Maybe rename as `register...`/`link....`/`bind...` could help 🤔

**Review by danielezonca — APPROVED** [review]
Just a minor comment

File: manager/src/main/java/com/redhat/developer/manager/ProcessorService.java (hunk 1 above)
Comment (r00ta) [reply manager/src/main/java/com/redhat/developer/manager/ProcessorService.java:49]: You sure @robpblake ? I've just tried and it works

Comment (robpblake) [reply manager/src/main/java/com/redhat/developer/manager/ProcessorService.java:49]: You make me doubt myself now. Let me double check.

Comment (robpblake) [reply manager/src/main/java/com/redhat/developer/manager/ProcessorService.java:49]: @r00ta Just to double check: you're suggesting I remove the `bridge.addProcessor(p);` line?

When I do that I get (as I would expect):

```
Caused by: org.h2.jdbc.JdbcSQLException: NULL not allowed for column "BRIDGE_ID"; SQL statement:
insert into Processor (bridge_id, name, published_at, status, submitted_at, version, id) values (?, ?, ?, ?, ?, ?, ?) [23502-197]
```

Nothing else is setting up the relationship between the entities other that the `addProcessor(p)` call.

File: manager/src/test/java/com/redhat/developer/manager/utils/TestUtils.java [hunk 5]
```diff
@@ -2,61 +2,58 @@
...
         return given()
                 .filter(new ResponseLoggingFilter())
                 .contentType(ContentType.JSON)
-                .when()
+                .when();
+    }
+
+    public static Response getBridges() {
+        return jsonRequest()
                 .get("/api/v1/bridges");
```
[the remark this answers is not in the document]
Comment (robpblake) [reply manager/src/test/java/com/redhat/developer/manager/utils/TestUtils.java:25]: 👍

File: manager/src/main/java/com/redhat/developer/manager/ProcessorService.java (hunk 1 above)
Comment (r00ta) [reply manager/src/main/java/com/redhat/developer/manager/ProcessorService.java:49]: Let me try again

Comment (r00ta) [reply manager/src/main/java/com/redhat/developer/manager/ProcessorService.java:49]: Sorry @robpblake , I mean to replace `bridge.addProcessor(p);` with `p.setBridge(bridge);`. In this very moment `bridge.addProcessor(p);` is setting adding the processor `p` to its list and the bridge `bridge` to the processor `p`. We can remove this if you simply do `p.setBridge(bridge);` here

**Comment by robpblake** [comment]
@r00ta / @danielezonca Updated. I decided to remove the bi-directional association between Bridge and Processor as thinking through the workflows we probably don't need it and it only really works if the number of Processors on a bridge remains low. This resolves the discussions around `addProcessor()/removeProcessor()` methods on the `Bridge`.

Factored in all other feedback from you both.

cheers,

Rob

**Review by r00ta — APPROVED** [review]
